src/main/flight/feedforward.c

```diff
@@ -73,85 +70,95 @@ FAST_CODE_NOINLINE float feedforwardApply(int axis, bool newRcFrame, feedforward
         const float rxRate = 1.0f / rxInterval;
         float setpointSpeed = (setpoint - prevSetpoint[axis]) * rxRate;
         float absPrevSetpointSpeed = fabsf(prevSetpointSpeed[axis]);
-        float setpointAcceleration = 0.0f;
-        const float ffSmoothFactor = pidGetFfSmoothFactor();
-        const float ffJitterFactor = pidGetFfJitterFactor();
+        float setpointAcceleration = setpointSpeed - prevSetpointSpeed[axis];
+        const float feedforwardSmoothFactor = pidGetFeedforwardSmoothFactor();
+        const float feedforwardJitterFactor = pidGetFeedforwardJitterFactor();
+        float feedforward;
 
         // calculate an attenuator from average of two most recent rcCommand deltas vs jitter threshold
-        float ffAttenuator = 1.0f;
-        if (ffJitterFactor) {
-            if (rcCommandDelta < ffJitterFactor) {
-                ffAttenuator = MAX(1.0f - ((rcCommandDelta + prevRcCommandDelta[axis]) / 2.0f) / ffJitterFactor, 0.0f);
-                ffAttenuator = 1.0f - ffAttenuator * ffAttenuator;
+        if (axis == FD_ROLL) {
+            DEBUG_SET(DEBUG_FEEDFORWARD, 3, lrintf(rcCommandDelta * 100.0f)); // rcCommand packet difference = steps of 50 mean 2000 RC steps
+        }
+        rcCommandDelta = fabsf(rcCommandDelta);
+        float jitterAttenuator = 1.0f;
+        if (feedforwardJitterFactor) {
+            if (rcCommandDelta < feedforwardJitterFactor) {
+                jitterAttenuator = MAX(1.0f - ((rcCommandDelta + prevRcCommandDelta[axis]) / 2.0f) / feedforwardJitterFactor, 0.0f);
+                jitterAttenuator = 1.0f - jitterAttenuator * jitterAttenuator;
             }
         }
 
+        const float setpointPercent = fabsf(setpoint) / feedforwardMaxRate[axis];
+        float absSetpointSpeed = fabsf(setpointSpeed); // unsmoothed for kick prevention
+
         // interpolate setpoint if necessary
         if (rcCommandDelta == 0.0f) {
-            if (prevDuplicatePacket[axis] == false && fabsf(setpoint) < 0.98f * ffMaxRate[axis]) {
-                // first duplicate after movement
-                // interpolate rawSetpoint by adding (speed + acceleration) * attenuator to previous setpoint
-                setpoint = prevSetpoint[axis] + (prevSetpointSpeed[axis] + prevAcceleration[axis]) * ffAttenuator * rxInterval;
-                // recalculate setpointSpeed and (later) acceleration from this new setpoint value
-                setpointSpeed = (setpoint - prevSetpoint[axis]) * rxRate;
+            if (prevDuplicatePacket[axis] == false) {
+                // first duplicate after movement, interpolate setpoint and use previous acceleration
+                // don't interpolate if sticks close to centre or max, interpolate jitter signals less than larger ones
+                if (setpointPercent > 0.02f && setpointPercent < 0.95f) {
```

Comment:
Curious about chosen edges: 0.02f and 0.95f (instead of 0.98f x ffMaxRate[axis])

Reply:
Thanks for the review.

The code here is concerned with interpolating 'duplicates', ie when there is no change in a received packet value.  The intent is to 'fill in' a duplicate that occurs in the middle of regular stick travel, so that feedforward will not get an exaggerated glitch from the apparent sudden deceleration.

At both centre and maximum stick positions, there are valid reasons to have 'duplicates', because the sticks often come to an abrupt halt.  We don't want to interpolate those kinds of duplicates.

`setpointPercent` of 0.98 is equal to 0.98 * ffMaxRate, and is used elsewhere in feedforward.c, so is calculated once only in the init part.  

The value of 0.95 vs 0.98 was changed to improve the behaviour when people with high expo settings.  Duplicates there won't be interpolated forward, and this is needed when the sticks hit max position.  Some people don't have really solid end-stops and there is jitter here.  Because of expo, small rcCommand jitter means quite bit setpoint jitter which would get amplified by interpolation unless interpolation is suppressed.  Hence the non-interpolation region was widened from 2% to 5%.   The up-side is greater stability at max stick position and there is little in the way of downside. Feedforward max rate limit blocks issues on the fast approach to max position, and on return from max not a lot happens in the first 5% of setpoint change.

The value of 0.02 was to improve behaviour of the sticks at centre.

The problem at centre stick positions is that if we interpolate random single packet width noise elements at centre stick, the feedforward element from a single step up, followed by the same value being held, will result in feedforward that is twice as tall and twice as wide, and will generate undesirable boost elements when we return to zero.  Hence any step change with a duplicate will get almost four times the feedforward.  The RC signal for high rate RC links typically becomes a kind of PWM signal around centre for smooth slow inputs, and they require integration to be useful.  The feedforward smoothing factor acts as a leaky integrator to make a meaningful signal from this PWM type data.  Inevitably, duplicates are not uncommon.  If interpolation is applied to those duplicates, the relative impact of the duplicate on the resulting signal is particularly bad. We need 'un-distorted' PWM type jitter in this case, and that means we mustn't interpolate the duplicates.

Not interpolation duplicates is also useful if springs are heavy and there is a sudden hard stop at centre, especially for low link rates.  Here the preceding travel can be fast, and the duplicate when the sticks stop can push a big feedforward element across centre, and we don't want that to happen.

The only downside of not interpolating forward around centre is that if there is a true duplicate in the +/- 2% of centre, it will cause a bigger feedforward glitch than it would if it was outside of the 2% range.  However, due to expo, setpoint change is typically smaller close to centre, so the impact of a true duplicate isn't as much of a problem as it otherwise would be.